File: js/math/fuelMath.js
// Pure fuel-consumption math -- no DOM, no browser APIs, no state. Mirrored
// 1:1 in python/paceometer_math/fuel_math.py as an independent test oracle
// (see tests/golden_vectors/fuel_math.json).
//
// The only fuel-economy data available per vehicle (fueleconomy.gov's bulk
// CSV) is two EPA test-cycle MPG figures -- city and highway -- each tied to
// a fixed test-cycle average speed, not a speed-swept curve. Confirmed
// directly against fueleconomy.gov's own test-schedule page: the FTP/UDDS
// city cycle averages FUEL_CITY_TEST_AVG_MPH, the HWFET highway cycle
// averages FUEL_HIGHWAY_TEST_AVG_MPH.
//
// Between those two anchors, gallons/mile is modeled as a straight line --
// a reasonable interpolant given only two real data points, though
// paceometer_lit_review's lit_review.md section 3 (Wang et al. 2008,
// Transportation Research Part D, peer-reviewed) finds the true minimum
// sits around 31-43mph, inside this range, not exactly at either anchor.
//
// Above FUEL_HIGHWAY_TEST_AVG_MPH, a straight line through the two anchors
// keeps *improving* mileage as speed rises, which is backwards -- most
// real speeding-over-the-limit driving happens entirely above this anchor,
// so that would tell a driver they saved gas by speeding, the opposite of
// this app's point (found and fixed 2026-08-03). Instead, above the
// highway anchor, gallons/mile grows quadratically off the highway-test
// rate. The v^2 *shape* is grounded in basic aerodynamics -- drag force is
// proportional to v^2 and the power needed to overcome it to v^3, so fuel
// burned per mile from drag alone scales with v^2 (see lit_review.md
// section 3's overview + section 7a's primary sources). The specific
// *coefficient* is only calibrated from a widely-cited rule of thumb (100
// km/h -> 110 km/h costs roughly 10% more fuel) that lit_review.md itself
// flags as secondary-sourced and "verify against a primary source before
// citing academically" -- treat it as same-ballpark, not exact, pending a
// tighter fit against Greene (1981) or Wang et al. (2008) (see
// docs/TODO.md).
export const FUEL_CITY_TEST_AVG_MPH = 21.2;
export const FUEL_HIGHWAY_TEST_AVG_MPH = 48.3;

const KMH_TO_MPH = 0.621371;
const RULE_OF_THUMB_SLOWER_KMH = 100;
const RULE_OF_THUMB_FASTER_KMH = 110;
const RULE_OF_THUMB_FUEL_INCREASE_FRACTION = 0.1;

// Solves for k in gal/mile(v) = highwayRate * (1 + k*(v - highwayAnchor)^2)
// such that the rule-of-thumb pair above lands ~10% apart, the same way
// zoneCeilingMph (paceMath.js) is solved algebraically rather than
// hardcoded, so this stays reproducible/adjustable if the rule of thumb is
// ever replaced with a tighter primary-sourced figure.
function solveAboveHighwayCoefficient() {
  const slowerMph = RULE_OF_THUMB_SLOWER_KMH * KMH_TO_MPH;
  const fasterMph = RULE_OF_THUMB_FASTER_KMH * KMH_TO_MPH;
  const a = (fasterMph - FUEL_HIGHWAY_TEST_AVG_MPH) ** 2;
  const b = (slowerMph - FUEL_HIGHWAY_TEST_AVG_MPH) ** 2;
  const targetRatio = 1 + RULE_OF_THUMB_FUEL_INCREASE_FRACTION;
  return (targetRatio - 1) / (a - targetRatio * b);
}

export const ABOVE_HIGHWAY_DRAG_COEFFICIENT = solveAboveHighwayCoefficient();

// Gallons burned per mile at a given speed, for a vehicle whose EPA city/
// highway MPG figures are known.
export function gallonsPerMile(mph, cityMpg, highwayMpg) {
  const cityRate = 1 / cityMpg;
  if (mph <= FUEL_CITY_TEST_AVG_MPH) return cityRate;

  const highwayRate = 1 / highwayMpg;
  if (mph <= FUEL_HIGHWAY_TEST_AVG_MPH) {
    const slope =
      (highwayRate - cityRate) / (FUEL_HIGHWAY_TEST_AVG_MPH - FUEL_CITY_TEST_AVG_MPH);
    return cityRate + slope * (mph - FUEL_CITY_TEST_AVG_MPH);
  }

  const overHighway = mph - FUEL_HIGHWAY_TEST_AVG_MPH;
  return highwayRate * (1 + ABOVE_HIGHWAY_DRAG_COEFFICIENT * overHighway ** 2);
}

File: python/paceometer_math/fuel_math.py
"""Pure fuel-consumption math -- a Python mirror of js/math/fuelMath.js.

Dev-only reference/test oracle: this module never runs inside the shipped
app -- see docs/CLAUDE.md and README.md for the app's actual (JS-only)
runtime architecture. See js/math/fuelMath.js for the full model rationale,
including why gallons/mile grows quadratically above the highway anchor
(found and fixed 2026-08-03: a straight line through just the two EPA
anchors keeps improving past the highway anchor, which is backwards) and
which parts of that model are literature-grounded vs. a provisional
rule-of-thumb calibration pending a tighter primary-sourced fit.
"""

FUEL_CITY_TEST_AVG_MPH = 21.2
FUEL_HIGHWAY_TEST_AVG_MPH = 48.3

KMH_TO_MPH = 0.621371
RULE_OF_THUMB_SLOWER_KMH = 100
RULE_OF_THUMB_FASTER_KMH = 110
RULE_OF_THUMB_FUEL_INCREASE_FRACTION = 0.1


def _solve_above_highway_coefficient():
    slower_mph = RULE_OF_THUMB_SLOWER_KMH * KMH_TO_MPH
    faster_mph = RULE_OF_THUMB_FASTER_KMH * KMH_TO_MPH
    a = (faster_mph - FUEL_HIGHWAY_TEST_AVG_MPH) ** 2
    b = (slower_mph - FUEL_HIGHWAY_TEST_AVG_MPH) ** 2
    target_ratio = 1 + RULE_OF_THUMB_FUEL_INCREASE_FRACTION
    return (target_ratio - 1) / (a - target_ratio * b)


ABOVE_HIGHWAY_DRAG_COEFFICIENT = _solve_above_highway_coefficient()


def gallons_per_mile(mph, city_mpg, highway_mpg):
    """Gallons burned per mile at a given speed, from EPA city/highway MPG."""
    city_rate = 1 / city_mpg
    if mph <= FUEL_CITY_TEST_AVG_MPH:
        return city_rate

    highway_rate = 1 / highway_mpg
    if mph <= FUEL_HIGHWAY_TEST_AVG_MPH:
        slope = (highway_rate - city_rate) / (FUEL_HIGHWAY_TEST_AVG_MPH - FUEL_CITY_TEST_AVG_MPH)
        return city_rate + slope * (mph - FUEL_CITY_TEST_AVG_MPH)

    over_highway = mph - FUEL_HIGHWAY_TEST_AVG_MPH
    return highway_rate * (1 + ABOVE_HIGHWAY_DRAG_COEFFICIENT * over_highway**2)
